### PR TITLE
fixes #312

### LINKED ISSUE #312: [Question/Bug] How to use fromJSON and defaultProps
### Version

3.2.24

### Expected behaviour

Allow use custom deserialize functions and the defaultProps

### Actual behaviour

the fromJSON function is not using the defaultProps from t.struct

### Steps to reproduce

```js
const t = require('tcomb');
const fromJSON = require('tcomb/lib/fromJSON');

t.Date.fromJSON = function(x) {
  return isNaN(Date.parse(x)) ? x : new Date(x);
};

const Sample = t.struct(
  {
    date: t.Date
  },
  {
    defaultProps: {
      get date() {
        return new Date();
      }
    }
  }
);

const sample = Sample({});
const json = JSON.parse(JSON.stringify(sample));
console.log(fromJSON(json, Sample));
console.log(fromJSON({}, Sample));
```

### Stack trace and console log

```
➔ node date.test.js
Struct { date: 2018-02-22T21:16:17.823Z }
/home/rkmax/my_project/node_modules/tcomb/lib/fail.js:2
  throw new TypeError('[tcomb] ' + message);
  ^

TypeError: [tcomb] Invalid value undefined supplied to Struct{date: Date}/date: Date
    at Function.fail (/home/rkmax/my_project/node_modules/tcomb/lib/fail.js:2:9)
    at assert (/home/rkmax/my_project/node_modules/tcomb/lib/assert.js:14:12)
    at Irreducible (/home/rkmax/my_project/node_modules/tcomb/lib/irreducible.js:18:7)
    at create (/home/rkmax/my_project/node_modules/tcomb/lib/create.js:9:104)
    at fromJSON (/home/rkmax/my_project/node_modules/tcomb/lib/fromJSON.js:26:12)
    at fromJSON (/home/rkmax/my_project/node_modules/tcomb/lib/fromJSON.js:61:20)
    at Object.<anonymous> (/home/rkmax/my_project/date.test.js:25:13)
    at Module._compile (module.js:643:30)
    at Object.Module._extensions..js (module.js:654:10)
    at Module.load (module.js:556:32)

```

### PR DESCRIPTION
**Pull request by rkmax**

### CODE REVIEW

File: lib/fromJSON.js [hunk 1]
```diff
@@ -57,8 +57,12 @@ function fromJSON(value, type, path) {
       var props = type.meta.props;
       ret = {};
       for (k in props) {
+        var actual = value[k];
+        if (actual === undefined) {
+          actual = type.meta.defaultProps[k];
```
Comment (gcanti) [inline lib/fromJSON.js]: maybe here we could save a reference to `defaultProps` for faster access

```js
var props = type.meta.props;
var defaultProps = type.meta.defaultProps;
...
  actual = defaultProps[k];
```

File: test/fromJSON.js [hunk 2]
```diff
@@ -77,8 +77,18 @@ describe('fromJSON', function () {
   it('should handle struct', function () {
     var MyType = t.struct({
       name: t.String,
-      birthDate: MyDate
-    }, 'MyType');
+      surname: t.String,
```
Comment (gcanti) [inline test/fromJSON.js]: could you please add a new specific test instead of modifying an existing one?

Comment (rkmax) [reply test/fromJSON.js]: sure

**Comment by gcanti** [comment]
Thanks @rkmax

**Pull request merged**